### ISSUE TITLE
Listen to specific value from StateNotifierProvider inside other provider

### ISSUE THREAD

**Issue by Nolence** [rrousselGit/river_pod#158]
I created a model of the app's state using `freezed`. This would include chats, user and other global info.

```dart
@freezed
abstract class AppState with _$AppState {
  const factory AppState({
    @nullable @required User user,
    @nullable @required AuthDetails authDetails,
    @nullable @required Chat chat,
    // ...
  }) = _AppState;
}
```

Then I made an `AppNotifier` that actually held the state of the app and managed it. I exposed it via a simple `StateNotifierProvider`.

```dart
class AppNotifier extends StateNotifier<AppState> {
  AppNotifier(this.ref) : super(AppState.empty());

  final ProviderReference ref;

  User get user => state.user;
  AuthDetails get authDetails => state.authDetails;
  Chat get chat => state.chat;

  Future<User> signIn() async {
    // ...
    final apiService = ref.read(apiProvider);
    final user = apiService.signIn();

    state = state.copyWith(user: user);

    return user;
  }
}

final appProvider = StateNotifierProvider<AppNotifier>((ref) {
  return AppNotifier(ref);
});
```

But now I want to create a userProvider that is a `FutureProvider` that will listen to ONLY the user on the `StateNotifier` and update when it changes. The reason it is a `FutureProvider` is that if it sees that the App's user is null, it will call `signIn` and return that user or null if they really can't sign in.

```dart
final userProvider = FutureProvider.autoDispose<User>((ref) async {
  final app = ref.watch(appProvider.state);

  if (app.user != null) {
    return Future.value(app.user);
  }

  try {
    final user = await ref.watch(appProvider).signIn();

    return user;
  } catch (_) {
    return Future.value(null);
  }
});
```

The problem is that there is no way to listen to a specific value on a StateNotifier and if any value gets changed (such as the `chat` in my example) userProvider will get called again forcing another loading state which is undesirable.

Is there a way to accomplish? To listen to individual values on a StateNotifier inside another provider? Something akin to

```dart
final userProvider = FutureProvider.autoDispose<User>((ref) async {
  final user = ref.watch(appProvider.state.select((state) => state.user));
  // ...
});
```

Thank you !

**Comment by iamarnas** [comment]
@Nolence Have you tried `AsyncValue`? 
```dart
class​ ​AppNotifier​ ​extends​ ​StateNotifier​<AsyncValue<​AppState​>>​ {
  ​AppNotifier​(​) ​:​ ​super​(const Asyncvalue.loading());
//... 
}
 ```

**Comment by Nolence** [comment]
I just checked it out with my AppNotifier but then that creates a host of other questions in how to manage state. For example, how would I go about actually making changes from inside the AppNotifier to the state? See the `signIn` method on AppNotifier that sets the value of state to include `user`.

I don't want to go too far in alternative solutions though before I understand if/why my current implementation is incorrect

**Comment by rrousselGit** [comment]
>   final user = ref.watch(appProvider.state.select((state) => state.user));


For now this is not supported. I will add support for that syntax in the near future.

An alternative is to extract that `state.user` in a `Provider`:

```dart
final _user = Provider((ref) {
   return ref.watch(appProvider.state).user;
});

final userProvider = FutureProvider.autoDispose<User>((ref) async {
  final user = ref.watch(_user);
  // ...
});
```

This will correctly filter undesired rebuilds.

**Comment by Nolence** [comment]
Thanks Remi !

That alternative works perfectly in the meantime

**Comment by tbm98** [comment]
> For now this is not supported. I will add support for that syntax in the near future.

🎉 🎉 🎉

**Comment by suragch** [comment]
I'm looking forward to select support. 

```
final user = ref.watch(appProvider.state.select((state) => state.user));
```

I believe this will also answer the question I asked [here](https://stackoverflow.com/questions/64346019/flutter-riverpod-statenotifierprovider-only-watch-changes-for-part-of-a-model).

**Comment by Marco87Developer** [comment]
> >   final user = ref.watch(appProvider.state.select((state) => state.user));
> 
> 
> For now this is not supported. I will add support for that syntax in the near future.
> 
> An alternative is to extract that `state.user` in a `Provider`:
> 
> ```dart
> final _user = Provider((ref) {
>    return ref.watch(appProvider.state).user;
> });
> 
> final userProvider = FutureProvider.autoDispose<User>((ref) async {
>   final user = ref.watch(_user);
>   // ...
> });
> ```
> 
> This will correctly filter undesired rebuilds.

It seems that those extra providers are not needed, as I show [here](https://stackoverflow.com/posts/comments/114152438?noredirect=1).

**Comment by rrousselGit** [comment]
They are.

The purpose of `_user` is that if anything _other_ then the `user` property updates, things that depends on the `_user` provider will not update

**Comment by Marco87Developer** [comment]
Thank you very much @rrousselGit for the answer! What I understand is the following. If we have:

```dart
final _user = Provider((ref) {
   return ref.watch(appProvider.state).user;
});

final userProvider = Provider<User>((ref) {
  return ref.watch(_user);
});
```

then we could say that `userProvider` is useless. The same would be with the following code:

```dart
final _buttonState = Provider<ButtonState>((ref) {
  return ref.watch(timerProvider.state).buttonState;
});

final buttonProvider = Provider<ButtonState>((ref) {
  return ref.watch(_buttonState);
}); 
```

The difference is that [here](https://github.com/rrousselGit/river_pod/issues/158#issuecomment-699638027) inside the `userProvider` we are using the `_user` to do something more than simply returning it, while in `buttonProvider` we are simply returning `_buttonState`, so that `buttonProvider` actually becomes useless (and the result using the `buttonProvider` will be the same as using `_buttonState`).

Did I get it right?

**Comment by rrousselGit** [comment]
Yes

On the other hand the private providers that you mentioned are useful and do reduce rebuilds

**Comment by Marco87Developer** [comment]
Thank you very much!

**Comment by venkatd** [comment]
@rrousselGit could explain the difference between calling `ref.watch(myProvider)` and `ref.watch(myProvider.state)`. It seems like you need to go with the 2nd option to trigger rebuilds both with `ref.watch` and `useProvider`. What is the reason for this?

**Comment by rrousselGit** [comment]
That is correct

This behavior simplifies dealing with buttons when the StateNotifier uses `autoDispose`, as `context.read` could lead to the notifier being disposed when still in use

`watch` is what keeps states alive.

**Comment by rrousselGit** [comment]
Closing as support for `ref.watch(provider.select)` landed in 1.0.0-dev
